plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -117,65 +136,99 @@ lensProvider
   CodeLensParams {_textDocument = TextDocumentIdentifier {_uri}}
     -- VSCode uses URIs instead of file paths
     -- haskell-lsp provides conversion functions
-    | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri _uri = liftIO $
+    | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri _uri =
       do
-        mbMinImports <- runAction "MinimalImports" state $ useWithStale MinimalImports nfp
+        mbMinImports <- liftIO $ runAction "MinimalImports" state $ useWithStale MinimalImports nfp
         case mbMinImports of
-          -- Implement the provider logic:
-          -- for every import, if it's lacking a explicit list, generate a code lens
-          Just (MinimalImportsResult minImports, posMapping) -> do
-            commands <-
-              sequence
-                [ generateLens pId _uri edit
-                  | (imp, Just minImport) <- minImports,
-                    Just edit <- [mkExplicitEdit pred posMapping imp minImport]
-                ]
-            return $ Right $ InL $ catMaybes commands
+          Just (MinimalImportsResult minImports _ _, posMapping) -> do
+            let lens = [ generateLens _uri curRange int
+                       | (range, int) <- minImports
+                       , let Just curRange = toCurrentRange posMapping range]
+            return $ Right $ InL lens
           _ ->
             return $ Right $ InL []
     | otherwise =
       return $ Right $ InL []
+  where generateLens :: Uri  -> Range -> Int -> CodeLens
+        generateLens uri range int =
+          CodeLens { _data_ = Just $ toJSON $ ResolveOne uri int
+                   , _range = range
+                   , _command = Nothing }
+
+lensResolveProvider :: PluginMethodHandler IdeState Method_CodeLensResolve
+lensResolveProvider ideState plId cl@(CodeLens {_data_ = Just data_}) = pluginResponse $ do
+  case fromJSON data_ of
```

Comment:
This applies in a few other places where you've done something like this, but if you like you can use `ViewPatterns` for something like this, e.g.
```
CodeLens {_data = Just (fromJSON -> Success (ResolveOne uri int)) } = ...
```
It's good for cases where you don't want to do different handling if the different parts fail. Arguably we might want to do that here: we might want to log differently if the code lens is missing a data field versus if the field is present but doesn't have the form we expect.